Add implementation for irregular dimension selection for nested tensors.

ani300 opened the pull request:
Summary: This diff modifies the implementation of the select operator so slices of the irregular dimension can be selected (e.g. nt[:,0,:]).

Test Plan:
Added new unit tests to test that the new functions work as intended (see them in diff). To test,
`buck test mode/dev-nosan //caffe2/test:nested`

Differential Revision: D41083993

Skylion007 commented on aten/src/ATen/native/nested/NestedTensorUtils.h:
```diff
@@ -320,17 +331,82 @@ inline Tensor wrap_tensor_node(
+    });
+    options = nt_buffer.options().merge_in(options_);
+  } else { // Slow path
+    std::vector<Tensor> flat_tensors;
+    std::vector<Tensor> sizes;
```
```suggestion
    std::vector<Tensor> flat_tensors;
    flat_tensors.reserve(tensor_node.degree());
    std::vector<Tensor> sizes;
    sizes.reserve(tensor_node.degree());
```

cpuhrsch commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -494,23 +494,60 @@ Tensor NestedTensor_sum_dim_CPU(
+        strides[positive_index],
+        offsets[positive_index]);
+  } else {
+    auto smallest_size = NestedTensor_get_min_size(*self_ptr);
```
Since you only need this value of bounds checking, why not do this as part of the loop below?

cpuhrsch commented on aten/src/ATen/native/nested/NestedTensorUtils.cpp:
```diff
@@ -57,6 +58,34 @@ std::vector<int64_t> NestedTensor_get_max_size(const NestedTensorImpl& nt) {
+  return results;
+}
+
+std::vector<int64_t> NestedTensor_get_min_size(const NestedTensorImpl& nt) {
```
Just before we forget: This is now unused :)

ani300 commented:
@pytorchbot merge -g

Commit:
Add irregular dimension selection for nested tensors. (#88585)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/88585

This diff modifies the implementation of the select operator so slices of the irregular dimension can be selected (e.g. nt[:,0,:]).

Test Plan:
Added new unit tests to test that the new function works as intended (see them in diff). To test,
`buck test mode/dev-nosan //caffe2/test:nested`

Reviewed By: cpuhrsch

Differential Revision: D41083993

fbshipit-source-id: 9666dc4e3ba7ef9e0edf3e3c655a1150eeaa5268

ani300 commented:
@pytorchbot merge -g
